src/modules/job-list/job_state.c

```diff
@@ -390,6 +391,34 @@ static const char *parse_job_name (const char *path)
     return path;
 }
 
+static int parse_per_resource (struct list_ctx *ctx,
+                               struct job *job,
+                               json_t *jobspec,
+                               const char **type,
+                               int *count)
+{
+    json_error_t error;
+
+    if (json_unpack_ex (jobspec, &error, 0,
+                        "{s:{s:{s?:{s?:{s?:{s?:s s?:i}}}}}}",
+                        "attributes",
+                        "system",
+                        "shell",
+                        "options",
+                        "per-resource",
+                        "type",
+                        type,
+                        "count",
+                        count) < 0) {
+        flux_log (ctx->h, LOG_ERR,
+                  "%s: job %ju invalid jobspec: %s",
+                  __FUNCTION__, (uintmax_t)job->id, error.text);
+        return -1;
+    }
+
+    return 0;
+}
```

Comment:
It feels like this is the sort of code that should be in a separate source module with unit tests.

Have you thought about whether it would be feasible to abstract the code that parses jobspec and R into a `struct job` into a  standalone and potentially reusable class or set of classes?

Reply:
i've thought about it on and off.  Lemme consider it more deeply with this PR and also considering #4336

Reply:
i already went down the path to do some split out in #4436, so perhaps its a good idea to try and do a split out now rather than later.  I'll try to do that in a separate PR.